### PR TITLE
Implement Amend Dialog

### LINKED ISSUE #94: "Amend commit" advanced dialog
Amending a commit in Git enables a developer to modify commit message, but can do much more. It can also alter author date, creator...

An advanced dialog to change those attributes would be, I think, a nice addition. It would be just a matter of displaying the dialog on triggering the "Amend commit" option. That'd be also a welcome break from the current behavior of the GUI which seems to be just cancelling the commit, therefore incorrect since this is what "Reset" is intended for

### PR DESCRIPTION
**Pull request by Murmele**
Currently amend is implemented as soft reset. This might lead to undesired changes lost if someone did not realized that it did a reset.

Implement a dialog and change data there and then amend with the libgit2 functionality.

fixes #94 and #265

### CODE REVIEW

**Comment by Murmele** [comment]
image

image

**Comment by Murmele** [comment]
@Minabsapi @lonix1 is this what you would like to have?

TODO: currently the date will be used from the old commit. Should it be changed too?

**Comment by Murmele** [comment]
@exactly-one-kas what do you think?

**Comment by Minabsapi** [comment]
> @Minabsapi @ lonix1 is this what you would like to have?

A supplementary field to edit author date would also be nice, but this aside, this is what I'd like

**Comment by Murmele** [comment]
> 

I think this is not the same thing as amend. Then you have to do a rebase soft or?

**Comment by Murmele** [comment]
> 

I think this is not the same thing as amend. Then you have to do a rebase soft or? At least from the libgit2 function it seems to be not possible doing something like this.

**Comment by Murmele** [comment]
> That is unfortunate, but I understand you about the complexity.
> 
> But, if that's the case, please rename it from amend? git amend and git soft reset are two different things, and it will be very confusing to people to do it and find it does something else...that's what happened to me and so I opened the original issue.

Why renaming it? It is amend withouth the possibility to add new files. (The new implementation using libgit2 git_commit_ammend())

**Comment by Murmele** [comment]
> 

I am curious what is for you the difference between soft reset + committing again and amend?

For me amend gives the possibility to change the properties of a commit.
- name, email for author committer (now possible)
- date of the original commit, date of the committer commit (not implemented in Gittyup)
- as I know now thanks to you it is possible to add new files (not possible in Gittyup)
```
# Edit hello.py and main.py
git add hello.py
git commit 
# Realize you forgot to add the changes from main.py 
git add main.py 
git commit --amend --no-edit
```
https://www.atlassian.com/git/tutorials/rewriting-history

A soft reset is a reset to the previous commit staging all the files again which have been in the commit. If I change the commit message and commit again then it is the same as I am doing an amend and changing only the commit message.

So one thing you can achive with amend what is not possible with a soft reset is that you add a file without changing the author name and email (Because when committing again, commiter and author got set to the current user).

The problem implementing amend with changing files is that we have to change the treeview (or adding it to the dialog) which makes it really complicated.

I think with this MR we have a correctly implemented amend command, with a limited set of capabilities (only name and and email can be changed).
Of course it is a big benefit, because now the amend command does not anything unexpected. And if you would like to edit a commit more than just editing credentials you just do a soft reset and committing afterwards.

**Comment by exactly-one-kas** [comment]
`amend` basically does the same as a soft reset + committing, albeit in one stroke (and therefore only writing one entry in the reflog)
`git_commit_amend` has the optional parameter `tree`
Using `git_index_write_tree` here to create a tree from the current index should achieve the desired effect of adding all staged changes to the amended commit

This has the same behavior as the CLI (and allows our users to have `amend` mean the same thing across all GIT frontends) and also allows changes to amend to be added externally

File: src/ui/RepoView.cpp [hunk 1]
```diff
@@ -2150,18 +2151,50 @@ void RepoView::promptToDeleteTag(const git::Reference &ref) {
   dialog->open();
 }
 
-void RepoView::promptToReset(const git::Commit &commit, git_reset_t type,
+void RepoView::promptToAmend(const git::Commit &commit,
                              const git::Commit &commitToAmend) {
+  // TODO: maybe also editing date...
```
Comment (exactly-one-kas) [inline src/ui/RepoView.cpp:2156]: Remove this - I created an issue for this: #277

**Review by exactly-one-kas — CHANGES_REQUESTED** [review]
LGTM, but I'd also prefer to have staged changes added to the amended commit (see https://github.com/Murmele/Gittyup/issues/273#issuecomment-1252779508)

**Comment by Murmele** [comment]
@exactly-one-kas thanks for the clarification. How would you like to implement it? Not sure how to do this graphically?

**Comment by Murmele** [comment]
> ```
> ...OOPS! I forgot some things
> git add blah.c
> git add qux.js
> git commit --amend   # uses same message
> # oops! I'm a dummy... fix blah.c again!
> git add blah.c
> git commit --amend   # uses same message
> ```
> 
> 
>     
>   
> 
> In most GUI tools all that can be done like this:
> 
> 
>     * stage files
> 
> 
>     * amend
> 
> 
> 
> 
> Incredibly easy and useful.

Ah that makes sense implementing like this. So you stage before and then you are doing the amend. I think this can be implemented. Sorry for my missunderstanding.

File: src/ui/RepoView.cpp (hunk 1 above)
Comment (Murmele) [inline src/ui/RepoView.cpp:2156]:
```suggestion
```

**Comment by Murmele** [comment]
> No, _I'm sorry_ for harassing you. 😄 I got a little excited because Gittyup has such a promising future. There are no modern cross platform gitkraken/sourcetree-like apps anymore, which are actively maintained. Except for Gittyup. It's really awesome. Thanks for maintaining it!

Yes this is true 😄 this was also the reason why I forked GitAhead a few years ago 😄 especially on linux there are not that many guis. GitKraken is just slow . Gittyup is not perfect there but way better 😄 Thanks to all the contributors especially to @exactly-one-kas i think we setup up a really good infrastructure to easly manage the project and pushing it forward 🚀

**Comment by Murmele** [comment]
https://github.com/libgit2/libgit2/blob/ac0f2245510f6c75db1b1e7af7ca01c15dec26bc/tests/libgit2/object/commit/commitstagedfile.c

Then lets implementing amending the staged files

**Comment by Murmele** [comment]
> https://github.com/libgit2/libgit2/blob/ac0f2245510f6c75db1b1e7af7ca01c15dec26bc/tests/libgit2/object/commit/commitstagedfile.c
> 
> Then lets implementing amending the staged files

Ok it was easier than expected to implement amending files

**Review by exactly-one-kas — APPROVED** [review]

**Comment by exactly-one-kas** [comment]
@Murmele I fixed a thing, can you look over it?

**Comment by Murmele** [comment]
Oh I did not realize that there is already an amendCommit Test haha ^^

Will this be still correctly?
image
image

Maybe we can change the message somehow that the parent commit is not needed?

**Comment by exactly-one-kas** [comment]
> Will this be still correctly?

How about just moving the message to after `mRepo.amend()` and  using `head.target().link()`?

**Comment by Murmele** [comment]
> 

Yes we can try

**Comment by Murmele** [comment]
> 

I tried it, but it seems to be not correct

**Comment by Murmele** [comment]
@exactly-one-kas I pushed another proposal

**Comment by exactly-one-kas** [comment]
Using the parent would simply be wrong, since the branch isn't pointing to the parent - it's pointing to the newly amended commit

**Comment by exactly-one-kas** [comment]
@Murmele does my proposal work?

**Comment by Murmele** [comment]
> > Will this be still correctly?
> 
> How about just moving the message to after `mRepo.amend()` and using `head.target().link()`?

I that case is is pointing to the commit id previously was used by the commit which will be amended

**Comment by Murmele** [comment]
This is how it looks now with by proposal:
Startpoint
image

After amend
image

**Comment by Murmele** [comment]
> @Murmele does my proposal work?

@exactly-one-kas  hmm sorry I did something wrong when I tested first time and then it showed me again the old id. using the new commit id is great.

**Comment by exactly-one-kas** [comment]
I'd say yes

**Comment by Murmele** [comment]
From my side this is ready

**Comment by Murmele** [comment]
@lonix1 thanks for the compliment. Great to hear that it simplifys your workflow

**Pull request merged**